### ISSUE TITLE
backward for dense+sparse does not work

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#43347]
## To Reproduce
```
a=torch.randn(2,3).to_sparse().requires_grad_(True)
c=torch.randn(2,3)
sum2=c+a
loss=sum2.sum()
loss.backward()
```
results in 
```
Function AddBackward0 returned an invalid gradient at index 1 - expected type TensorOptions(dtype=float, device=cpu, layout=Sparse, requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt)) but got TensorOptions(dtype=float, device=cpu, layout=Strided, requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))
Exception raised from validate_outputs at caffe2/torch/csrc/autograd/engine.cpp:622 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x7b (0x7fb71562896b in /mnt/xarfuse/uid-202799/443637f6-seed-nspid4026531836-ns-4026531840/libomnibus.so)
frame #1: torch::autograd::validate_outputs(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> >&, std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&) + 0x13db (0x7fb705c28cdb in /mnt/xarfuse/uid-202799/443637f6-seed-nspid4026531836-ns-4026531840/libomnibus.so)
frame #2: <unknown function> + 0x7d8eb186 (0x7fb705c2a186 in /mnt/xarfuse/uid-202799/443637f6-seed-nspid4026531836-ns-4026531840/libomnibus.so)
```

**Comment by ngimel** [comment]
To fix this, we need to decide what the semantics of dense+sparse operation is. Is it more like index_add, where the values from sparse tensor are added at particular indices of dense, or is it like regular addition, where sparse tensor is just a shorthand for a tensor that has a lot of zeros? In the former case, the gradient for sparse tensor should be a tensor with the same non-zero indices as original, propagating gradOutput values at the corresponding indices, in the latter the gradient for the sparse tensor will, in fact, be dense gradOutput, that we can nominally convert to sparse, but all the elements will be nonzeros, so it will take more memory because all the indices will need to be stored along with values.
I think if people want `index_add` semantics they should be using `index_add` (`index_sub`, `index_mul` etc), and we should add those functions and make it so that they can directly take indices and values from sparse tensors, but it is possible that users are expecting `index_add` semantics when writing `a+b`.

**Comment by albanD** [comment]
I would look at it slightly differently and follow the same logic we have about device, type, etc: the gradient Tensor should be of same size/device/type as the Tensor used in forward unless explicitly asked for by the user.

The current explicit cases we have right now are embedding(sparse=True) and mixed precision training. 
And so I would lean towards the index_add idea you presented: a sparse Tensor in forward means a sparse Tensor in backward.

What do you think?

**Comment by ngimel** [comment]
I agree that sparse tensor in forward means sparse tensor in backward, my question is, how many nonzero values should that sparse tensor have. A very quick patch doing what we currently do for device/dtype
```
diff --git a/torch/csrc/autograd/engine.cpp b/torch/csrc/autograd/engine.cpp
index dbfad76c05..c5d366dff7 100644
--- a/torch/csrc/autograd/engine.cpp
+++ b/torch/csrc/autograd/engine.cpp
@@ -546,8 +546,10 @@ static variable_list call_post_hooks(Node& fn, variable_list outputs, const vari
 
 static bool is_compatible_type(const at::TensorOptions& expected, const at::TensorOptions& actual) {
   // Types are compatible if they exactly match or if the gradient is a sparse
-  // version of the expected type.
-  return expected.type_equal(actual) || (actual.is_sparse() && expected.device().type() == actual.device().type());
+  // version of the expected type, or the gradient dense and expected is sparse
+
+  return expected.type_equal(actual) || ((actual.is_sparse() ||
+   (!actual.is_sparse() && expected.is_sparse())) && expected.device().type() == actual.device().type());
 }
 
 void set_device(int device) {
@@ -621,6 +623,9 @@ void validate_outputs(
        ss << metadata.options() << " but got " << grad.options();
        AT_ERROR(format_error(ss.str()));
     }
+    if (!grad.options().is_sparse() && metadata.options().is_sparse()) {
+      grad = grad.to_sparse();
+    }
     auto grad_device = grad.device();
     if (grad_device != metadata.device()) {
       std::stringstream ss;
```
makes the case above not error out, but also produces sparse gradient where nnz==numel (sorry, ipython printout because that's what I have):
```
In [2]: indices = torch.tensor([[0,1],[1,2]])                                                                                                                                                                      

In [3]: values=torch.tensor([2.,3.])                                                                                                                                                                               

In [4]: t=torch.sparse_coo_tensor(indices, values)                                                                                                                                                                 

In [5]: t                                                                                                                                                                                                          
Out[5]: 
tensor(indices=tensor([[0, 1],
                       [1, 2]]),
       values=tensor([2., 3.]),
       size=(2, 3), nnz=2, layout=torch.sparse_coo)

In [6]: b=torch.randn(2,3)                                                                                                                                                                                         

In [7]: td=t.to_dense().requires_grad_(True)                                                                                                                                                                       

In [8]: t.requires_grad_(True)                                                                                                                                                                                     
Out[8]: 
tensor(indices=tensor([[0, 1],
                       [1, 2]]),
       values=tensor([2., 3.]),
       size=(2, 3), nnz=2, layout=torch.sparse_coo, requires_grad=True)

In [9]: (b+t).sum().backward()                                                                                                                                                                                     

In [10]: (b+td).sum().backward()                                                                                                                                                                                   

In [11]: t.grad                                                                                                                                                                                                    
Out[11]: 
tensor(indices=tensor([[0, 0, 0, 1, 1, 1],
                       [0, 1, 2, 0, 1, 2]]),
       values=tensor([1., 1., 1., 1., 1., 1.]),
       size=(2, 3), nnz=6, layout=torch.sparse_coo)

In [12]: td.grad                                                                                                                                                                                                   
Out[12]: 
tensor([[1., 1., 1.],
        [1., 1., 1.]])
```
Here, gradient of sparse tensor is what gradient of dense tensor would be, just represented as sparse. If we followed index_add semantics, `grad` would have 2 nonzero elements, the same number of non-zeros sparse input had.

**Comment by albanD** [comment]
Ho ok! I missed your point earlier. Thanks for the clarifying example.
In index add, we ignore the gradient for all the entries that are not provided in values. While with sparse, even though the input contain many zero values, they actually do have some gradients.
Mathematically speaking, you expect to get a gradient that contain all these values. It would be wrong to force them to be 0 indeed (as you don't get the correct gradients anymore).

But thinking about it, for most operations, a sparse input (with large number of 0s) does not mean that the gradient will have many 0s. So there is no real gain at trying to propagate a sparse Tensor as a Dense Tensor is a better way to represent the gradients.

**Comment by pearu** [comment]
Notice that the memory issue of the gradient of sparse tensor being a full sparse tensor (containing non-zeros but all equal) would be resolved by introducing fill value to sparse tensors as in https://github.com/pytorch/rfcs/pull/8 because then we would have:
```
In [11]: t.grad                                                                                                                                                                                                    
Out[11]: 
tensor(indices=tensor([[], []]),
       values=tensor([]),
       fill_value=tensor(1.),
       size=(2, 3), nnz=6, layout=torch.sparse_coo)
```

Re "we need to decide what the semantics of dense+sparse operation is.": both `index_add` and regular addition can be supported when the user specifies `fill_value=<indefinite value>` (where the `<indefinite value>` placeholder needs to be decided) and `fill_value=0`, respecitvely.

**Comment by ngimel** [comment]
In the general case, the memory issue of the gradient won't be solved. In this particular case the gradOutput that sum operation received is all ones, but that's because this is synthetic case. In general, gradOutput will be a dense tensor with all different values.
